src/Workspaces/Core/Portable/Utilities/BKTree.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            _allLowerCaseCharacters = allLowerCaseCharacters;
+            _nodes = nodes;
+            _edges = edges;
+        }
+
+        public static BKTree Create(params string[] values)
+        {
+            return Create((IEnumerable<string>)values);
+        }
+
+        public static BKTree Create(IEnumerable<string> values)
+        {
+            return new Builder(values).Create();
+        }
+
+        public IList<string> Find(string value, int? threshold = null)
+        {
+            if (_nodes.Length == 0)
+            {
+                return SpecializedCollections.EmptyList<string>();
+            }
+
+            var lowerCaseCharacters = ArrayPool<char>.GetArray(value.Length);
+            try
+            {
+                for (var i = 0; i < value.Length; i++)
+                {
+                    lowerCaseCharacters[i] = char.ToLower(value[i]);
+                }
+
+                threshold = threshold ?? WordSimilarityChecker.GetThreshold(value);
+                var result = new List<string>();
+                Lookup(_nodes[0], lowerCaseCharacters, value.Length, threshold.Value, result);
+                return result;
+            }
+            finally
+            {
+                ArrayPool<char>.ReleaseArray(lowerCaseCharacters);
+            }
+        }
+
+        private void Lookup(Node currentNode, char[] queryCharacters, int queryLength, int threshold, List<string> result)
+        {
+            // We always want to compute the real edit distance (ignoring any thresholds).  This is
+            // because we need that edit distance to appropriately determine which edges to walk 
+            // in the tree.
+            var characterSpan = currentNode.CharacterSpan;
+            var editDistance = EditDistance.GetEditDistance(
+                new ArraySlice<char>(_allLowerCaseCharacters, characterSpan),
+                new ArraySlice<char>(queryCharacters, 0, queryLength));
+
+            if (editDistance <= threshold)
+            {
+                // Found a match.
+                result.Add(new string(_allLowerCaseCharacters, characterSpan.Start, characterSpan.Length));
+            }
+
+            var min = editDistance - threshold;
+            var max = editDistance + threshold;
+
+            var firstEdgeIndex = currentNode.FirstEdgeIndex;
+            var lastEdgeIndex = firstEdgeIndex + currentNode.EdgeCount;
+            for (var i = firstEdgeIndex; i < lastEdgeIndex; i++)
+            {
+                var childEditDistance = _edges[i].EditDistance;
+                if (min <= childEditDistance && childEditDistance <= max)
+                {
+                    Lookup(this._nodes[_edges[i].ChildNodeIndex],
+                        queryCharacters, queryLength, threshold, result);
+                }
+            }
+        }
+
+        internal void DumpStats()
```

Comment:
Did you mean to include this? If you want to keep it, at least comment it as being for diagnostic purposes.